### ISSUE TITLE
Investigate Utf8JsonWriter performance improvement by using AggressiveInlining

### ISSUE THREAD

**Issue by ahsonkhan** [dotnet/corefx#34568]
From https://github.com/dotnet/corefx/pull/34425#discussion_r247299987

Given we do not pass spans by ref within the `Utf8JsonWriter`, it is likely fruitful to mark certain Write methods as `AggressiveInlining` to reduce the amount of struct copies.

Investigate the performance of deep call graphs and consider marking the common path as `AggressiveInlining` (if it shows noticeable performance improvement).

For example, something like this:
```C#
// => denotes a method call in this case (not valid C# syntax)
WriteStringValue(ReadOnlySpan<char> value, bool suppressEscaping = false)
=> WriteStringSuppressFalse(ReadOnlySpan<char> value)
=> WriteStringByOptions(ReadOnlySpan<char> value)
=> WriteStringMinimized(ReadOnlySpan<char> escapedValue)
=> WriteStringValue(ReadOnlySpan<char> escapedValue, ref int idx)
```

The above call graph includes `WriteStringSuppressFalse` which has just a single callsite, so it would ok to inline it into the caller - either using AggresiveInlining (preferred); or by just not having the separate method at all and just inlining the code in manually.

**Comment by omariom** [comment]
> Given we do not pass spans by ref within the Utf8JsonWriter,

Looks like  passing spans by value and by ref  now [generates identical code](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wIDoAlAVwDs4BLVAU2oGEB7VABxYAbdgCcAymIBuLAMbtkpMuUwBmfACZcXXOQDe5XEfzrMUXAAUUyAEIBPAGqIhACgCUussdwGv347TsiAAmAPKMQnbi/IiMADzAdnDsAHy4icnIuAC8uIzsAO7pSewA2gC6PrgEMBow6gC+Sv7G9k6uGQpuzcYNhsb9Rmr45lbItnaBAGbunt6+LUaBIeGR0bEJJWmdWbn5RZ0VVTV1jT0t9tMuO92DuH1+RnelALLscAAWPMEAkgKub0+3z+/CEoX4rB4jGQ1AAcjwfhEWIxkQBzNzlO78UQsKSIZJUCC4ZFwXBtZwuZZhCJRGLxTppZB0tx3BaLTAAdlwTNi1AAMuxGKjPud7ndnoCvr9/i5JcD/uDIdC4QikSihRisTi8QTCESSWTJuwZsjcFTVrSNgzuczWXdvJybbyBUKRXcHg0gA===).

**Comment by ahsonkhan** [comment]
> Looks like  passing spans by value and by ref  now [generates identical code]

From the disassembly you shared, PassByVal has one extra `lea` instruction:
https://www.diffchecker.com/68FWHi0b

**Comment by Tornhoof** [comment]
My experience regarding ``in`` or not ``in`` for ReadOnlySpan is to be extremely careful if I don't use it, so I basically slap it on every ``ReadOnlySpan<byte>`` method parameter:
Take the following code snippet
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static ulong ReadUInt64(in ReadOnlySpan<byte> span, int offset)
{
    return Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref MemoryMarshal.GetReference(span), offset));
}
```
(https://github.com/Tornhoof/SpanJson/blob/master/SpanJson/Helpers/SpanHelper.cs)

it is heavily used for the integer based property matching during deserialization, the same code exists for byte, ushort, uint32 and is used for both UTF8 and UTF16 , the difference for UTF16 that prior to that snippet there is a ``MemoryMarshal.AsBytes`` to get the bytes instead of chars for the matching and there are more calls to it as there are twice the bytes for utf16.

If I remove the ``in`` from these methods (and do not change anything else), there is a 15% regression in performance in my nested serializer tests for UTF16 and only UTF16:

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i9-9900K CPU 3.60GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=3.0.100-preview-009844
  [Host]     : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT
  Job-MNYSVC : .NET Core 3.0.0-preview-27218-01 (CoreCLR 4.6.27217.02, CoreFX 4.7.18.61304), 64bit RyuJIT

UnrollFactor=2  

```

**With in**

|                                      Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|-------------------------------------------- |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|     DeserializeAnswerWithSpanJsonSerializer | 3.556 us | 0.0209 us | 0.0196 us |      0.2937 |      0.0038 |           - |             1.84 KB |
| DeserializeAnswerWithSpanJsonSerializerUtf8 | 4.129 us | 0.0149 us | 0.0139 us |      0.2899 |           - |           - |             1.84 KB |                                      

**Without In**									  

|         							   Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|-------------------------------------------- |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|     DeserializeAnswerWithSpanJsonSerializer | 4.081 us | 0.0462 us | 0.0432 us |      0.2899 |           - |           - |             1.84 KB |
| DeserializeAnswerWithSpanJsonSerializerUtf8 | 4.151 us | 0.0343 us | 0.0287 us |      0.2899 |           - |           - |             1.84 KB |

**Comment by Tornhoof** [comment]
> Also, this is very unsafe method since it avoids bounds-checks. You are not really getting any benefits from Span. You can as well just use pointers to make it even leaner and explicitly unsafe.
Yes I'm aware of that, but it's the only way to get it working with Expression Trees, as there is no ``byref`` support for Expression Trees and even viewing the generated IL from expression trees is not trivial anymore in .NET Core. But you say that it should not happen, I'll see if I can produce a repro which is smaller and file a bug.
